core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -892,3 +871,220 @@ object ZkPath {
     client.createPersistentSequential(path, data)
   }
 }
+
+/**
+ * Implements an ephemeral znode that is recreated in the case it disappears.
+ * A znode can be deleted in this case if it has been created in a previous session.
+ * Say that the client that created it crashes leaving the session open. In this case,
+ * this client will try to create it and fail, but eventually the ephemeral znode will
+ * be remove and needs to be created and associated to this session.
+ *
+ * It additionally creates the parent path recursively.
+ */
+class ZKWatchedEphemeral(path : String,
+                          data : String,
+                          expectedCallerData: Any,
+                          checker: (String, Any) => Boolean,
```

Comment:
It is basically leads to unsafe casts. You can see a number of them in the callers. With a type parameter, the casts are not needed and the compiler will ensure that the right type is used.